workflow/rules/samtools.smk

```diff
@@ -0,0 +1,137 @@
+    params:
+        extra=config.get("samtools_index", {}).get("extra", ""),
+    log:
+        "{file}.bam.bai.log",
+    benchmark:
+        repeat(
+            "{file}.bam.bai.benchmark.tsv",
+            config.get("samtools_index", {}).get("benchmark_repeats", 1),
+        )
+    container:
+        config.get("samtools_index", {}).get("container", config["default_container"])
+    threads: config.get("samtools_index", {}).get("threads", config["default_resources"]["threads"])
+    resources:
+        threads=config.get("samtools_index", {}).get("threads", config["default_resources"]["threads"]),
+        time=config.get("samtools_index", {}).get("time", config["default_resources"]["time"]),
+        mem_mb=config.get("samtools_index", {}).get("mem_mb", config["default_resources"]["mem_mb"]),
+        mem_per_cpu=config.get("samtools_index", {}).get("mem_per_cpu", config["default_resources"]["mem_per_cpu"]),
+        partition=config.get("samtools_index", {}).get("partition", config["default_resources"]["partition"]),
+    conda:
+        "../envs/samtools.yaml"
+    message:
+        "{rule}: create index for {wildcards.file}"
+    wrapper:
+        "0.78.0/bio/samtools/index"
+
+
+rule samtools_merge_bam:
+    input:
+        bams=expand(
+            "alignment/picard_mark_duplicates/{{sample}}_{{type}}_{chr}.bam",
+            chr=extract_chr(
+                "%s.fai" % (config["reference"]["fasta"]), filter_out=config.get("reference", {}).get("skip_chrs", [])
+            ),
+        ),
+    output:
+        bam=temp("alignment/samtools_merge_bam/{sample}_{type}.bam_unsorted"),
+    params:
+        extra=config.get("samtools_merge_bam", {}).get("extra", ""),
+    log:
+        "alignment/samtools_merge_bam/{sample}_{type}.bam_unsorted.log",
+    benchmark:
+        repeat(
+            "alignment/samtools_merge_bam/{sample}_{type}.bam_unsorted.benchmark.tsv",
+            config.get("samtools_merge_bam", {}).get("benchmark_repeats", 1),
+        )
+    threads: config.get("samtools_merge_bam", {}).get("threads", config["default_resources"]["threads"])
+    resources:
+        threads=config.get("samtools_merge_bam", {}).get("threads", config["default_resources"]["threads"]),
+        time=config.get("samtools_merge_bam", {}).get("time", config["default_resources"]["time"]),
+        mem_mb=config.get("samtools_merge_bam", {}).get("mem_mb", config["default_resources"]["mem_mb"]),
+        mem_per_cpu=config.get("samtools_merge_bam", {}).get("mem_per_cpu", config["default_resources"]["mem_per_cpu"]),
+        partition=config.get("samtools_merge_bam", {}).get("partition", config["default_resources"]["partition"]),
+    container:
+        config.get("samtools", {}).get("container", config["default_container"])
+    conda:
+        "../envs/samtools.yaml"
+    message:
+        "{rule}: Merge chr bam files, creating {output}"
+    shell:
+        "(samtools merge -c -p {params.extra} {output} {input}) &> {log}"
```

Comment:
there is a wrapper, not sure if it fits your need

Reply:
changed to it now